### PR TITLE
Implementation of sftp storage.

### PR DESCRIPTION
**Pull request by Mun1z**
The sftp protocol storage was implemented, but the unit and integration tests will be added in a different pull request.

### CODE REVIEW

File: ml_git/constants.py [hunk 1]
```diff
@@ -71,6 +72,8 @@ class StoreType(Enum):
     AZUREBLOBH = 'azureblobh'
     GDRIVEH = 'gdriveh'
     GDRIVE = 'gdrive'
+    SFTPH = 'sftph'
+    SFTP = 'sftp'
```
Comment (lucasncabral) [inline ml_git/constants.py:76]: Is it necessary to define the `sftp` type? Currently the implementation only supports the `sftph` (with the files breaking).

File: ml_git/storages/sftp_store.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+from ml_git.constants import STORE_FACTORY_CLASS_NAME, SFTPSTORE_NAME
+from ml_git.ml_git_message import output_messages
+from ml_git.storages.store import Store
+
+import paramiko
+
+
+class SFtpStore(Store):
+    def __init__(self, bucket_name, bucket):
+        self._store_type = 'SFTP'
```
Comment (lucasncabral) [inline ml_git/storages/sftp_store.py:18]: Why not use the value defined in the enum? (`StoreType.SFTPH.value`)

File: ml_git/storages/sftp_store.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        self._host = get_key('endpoint-url', bucket)
+
+        self._bucket = bucket_name
+        super(SFtpStore, self).__init__()
+
+    def connect(self):
+        ssh_client = paramiko.SSHClient()
+        ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+
+        mykey = paramiko.RSAKey.from_private_key_file(self._key)
```
Comment (lucasncabral) [inline ml_git/storages/sftp_store.py:30]: Please rename `mykey` variable to a more intuitive name (`user_key`?).

File: ml_git/storages/sftp_store.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        try:
+            self._store.chdir(self._bucket)
+        except IOError:
+            error_msg = output_messages['ERROR_BUCKET_DOES_NOT_EXIST'] % self._bucket
+            log.error(error_msg, class_name=STORE_FACTORY_CLASS_NAME)
+            return False
+
+        return True
+
+    def create_bucket_name(self, bucket_prefix):
```
Comment (lucasncabral) [inline ml_git/storages/sftp_store.py:49]: This method is duplicated (`storages/s3store.create_bucket_name`). I think the ideal would be to move it to the `storages/store_utils.py` file.

File: ml_git/storages/sftp_store.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+    def put(self, key_path, file_path):
+        bucket = self._bucket
+
+        with open(file_path, 'rb') as text_file:
+            self._store.put(file_path, os.path.basename(text_file.name))
+
+        version = None
+
+        log.info('Put - stored [%s] in bucket [%s] with key [%s]-[%s]' % (file_path, bucket, key_path, version),
```
Comment (lucasncabral) [inline ml_git/storages/sftp_store.py:75]: Please place this message in the `ml_git_messages.py` file.

File: ml_git/storages/sftp_store.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            return uri, None
+
+        version = remain[len(v):]
+        return key, version
+
+    def get(self, file_path, reference):
+        key, version = self._to_file(reference)
+        return self._get(file_path, key, version=version)
+
+    def get_object(self, key_path):
```
Comment (lucasncabral) [inline ml_git/storages/sftp_store.py:102]: I couldn't understand the logic of this method. Why do we need a 'try' block to throw an exception? And why catch an exception of a different type than the one that was thrown? Is this code block really necessary? 

Apparently it is always returning the exception that the object was not found.

File: ml_git/storages/sftp_store.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        try:
+            raise RuntimeError('Object [%s] not found' % key_path)
+        except IOError:
+            res = self._store.open(key_path)
+            return res.read(res.stat().st_size)
+
+    def _get(self, file, key_path, version=None):
```
Comment (lucasncabral) [inline ml_git/storages/sftp_store.py:109]: The same comment made above.

**Review by lucasncabral — CHANGES_REQUESTED** [review]
Please take a look at the commented points. In addition, add documentation related to this new type of storage in docs/storage_configurations.md. Also update the documentation of the commands that were affected (storage add, storage del, create).

File: ml_git/storages/sftp_store.py (hunk 7 above)
Comment (Mun1z) [reply ml_git/storages/sftp_store.py:110]: Resolved on commit: [6354d28](https://github.com/HPInc/ml-git/pull/68/commits/6354d28410904663ae41f8028037197034fa34f5)

File: ml_git/storages/sftp_store.py (hunk 6 above)
Comment (Mun1z) [reply ml_git/storages/sftp_store.py:102]: Resolved on commit: [6354d28](https://github.com/HPInc/ml-git/pull/68/commits/6354d28410904663ae41f8028037197034fa34f5)

File: ml_git/storages/sftp_store.py (hunk 5 above)
Comment (Mun1z) [reply ml_git/storages/sftp_store.py]: Resolved on commit: [6354d28](https://github.com/HPInc/ml-git/pull/68/commits/6354d28410904663ae41f8028037197034fa34f5)

File: ml_git/storages/sftp_store.py (hunk 4 above)
Comment (Mun1z) [reply ml_git/storages/sftp_store.py:49]: This generate a circular dependency.

File: ml_git/storages/sftp_store.py (hunk 3 above)
Comment (Mun1z) [reply ml_git/storages/sftp_store.py]: Resolved on commit: [6354d28](https://github.com/HPInc/ml-git/pull/68/commits/6354d28410904663ae41f8028037197034fa34f5)

File: ml_git/storages/sftp_store.py (hunk 2 above)
Comment (Mun1z) [reply ml_git/storages/sftp_store.py]: Resolved on commit: [6354d28](https://github.com/HPInc/ml-git/pull/68/commits/6354d28410904663ae41f8028037197034fa34f5)

File: ml_git/constants.py (hunk 1 above)
Comment (Mun1z) [reply ml_git/constants.py:76]: Resolved on commit: [6354d28](https://github.com/HPInc/ml-git/pull/68/commits/6354d28410904663ae41f8028037197034fa34f5)

**Review by lucasncabral — APPROVED** [review]

**Review by GustavoMaciel — APPROVED** [review]

**Review by gilMars — APPROVED** [review]

**Review by andersonpablosilva — CHANGES_REQUESTED** [review]

File: ml_git/admin.py [hunk 8]
```diff
@@ -93,7 +93,7 @@ def valid_store_type(store_type):
     return True
 
 
-def store_add(store_type, bucket, credentials_profile, global_conf=False, endpoint_url=None):
+def store_add(store_type, bucket, credentials_profile, global_conf=False, endpoint_url=None, username=None, ssh_key=None):
```
[the remark this answers is not in the document]
Comment (Mun1z) [reply ml_git/admin.py]: Done on commit [937ebcb](https://github.com/HPInc/ml-git/pull/68/commits/937ebcb6ef4486a8470ad33477acbf855039e533)

**Review by andersonpablosilva — APPROVED** [review]

File: tests/integration/fake_ssh_key/id_rsa [hunk 9]
```diff
@@ -0,0 +1,49 @@
+-----BEGIN OPENSSH PRIVATE KEY-----
```
Comment (diego-pedroso) [inline tests/integration/fake_ssh_key/id_rsa:1]: Can't we generate this key using CLI before the test execution?

File: docs/sftp_configurations.md [hunk 10]
```diff
@@ -0,0 +1,19 @@
+# SFTP bucket configuration #
+
+This section explains how to configure the settings that the ml-git uses to interact with your bucket using the SFTP storage. This requires you to have configured a public key in your SFTP server and use the private key pair to connect through ml-git.
```
Comment (diego-pedroso) [inline docs/sftp_configurations.md:3]: Is 'bucket' also a terminology used on SFTP servers, or are we using 'bucket' to define repositories/directories on all storage services?

Comment (Mun1z) [reply docs/sftp_configurations.md:3]: In this case we are using it to define the repositories/directories of any service.

File: tests/integration/fake_ssh_key/id_rsa (hunk 9 above)
Comment (Mun1z) [reply tests/integration/fake_ssh_key/id_rsa]: Yes, the change was made to generate the key via the command line during the execution of the tests.

**Review by diego-pedroso — APPROVED** [review]

**Pull request merged**